modules/assim.sequential/R/downscale_function_hrly.R

```diff
@@ -0,0 +1,108 @@
+#' 
+#' @author Harunobu Ishii
+#' @param nc_file  In quotes, file path for .nc containing ensemble data.
+#' @param coords In quotes, file path for .csv file containing the site coordinates, columns named "lon" and "lat".
+#' @param yyyy In string, format is yyyy(year of interest)
+#' @param covariates SpatRaster stack, used as predictors in randomForest. Layers within stack should be named. Recommended that this stack be generated using 'covariates' instructions in assim.sequential/inst folder
+#' @return It returns the `downscale_output` list containing lists for the training and testing data sets, models, and predicted maps for each ensemble member.
+#' @import ncdf4
+#' @import lubridate
+#' @export
+
+SDA_downscale_hrly <- function(nc_file, coords, yyyy, covariates){
+  
+  # Read the input data and site coordinates
+  nc_data <- nc_open(nc_file)
+  input_data <- ncvar_get(nc_data, "NEE")
+  covariate_names <- names(covariates)
+  
+  
+  # Extract time and units
+  time <- nc_data$dim$time$vals
+  time_units <- nc_data$dim$time$units
+  time_origin_str <- substr(time_units, 12, 31)
+  
+  # Check if timezone is specified in the time units string
+  if (grepl("UTC|GMT", time_units)) {
+    time_origin <- ymd_hm(time_origin_str, tz = "UTC")
+  } else if (grepl("EST", time_units)) {
+    time_origin <- ymd_hm(time_origin_str, tz = "EST")
+  } else {
+    time_origin <- ymd_hm(time_origin_str, tz = "UTC")  # Default to UTC if not specified
+  }
+  
+  # Timereadable
+  if (grepl("hours", time_units)) {
+    time_readable <- time_origin + dhours(time)
+  } else if (grepl("seconds", time_units)) {
+    time_readable <- time_origin + dseconds(time)
+  } else {
+    stop("Unsupported time units")
+  }
+  
+  # Extract predictors from covariates raster using site coordinates
+  site_coordinates <- terra::vect(readr::read_csv(coords), geom=c("lon", "lat"), crs="EPSG:4326")
+  predictors <- as.data.frame(terra::extract(covariates, site_coordinates,ID = FALSE)) 
+
+  downscale_output<- list()
+  
+  # Train & Test split
+  sample <- sample(1:nrow(predictors), size = round(0.75*nrow(predictors)))
+  
+  # Predict for each time stamp of the year selected
+  time_indices <- which(year(time_readable) == yyyy)
+  for (index in time_indices) {
+    if(index == 37986){
+      break
+    }
+    data <- input_data[index, , ]
+    carbon_data <- as.data.frame(data)
+    names(carbon_data) <- paste0("ensemble",seq(1:ncol(carbon_data)))
+
+    # Combine carbon data and covariates/predictors and split into training/test
+    full_data <- cbind(carbon_data, predictors)
+    train_data <- full_data[sample, ]
+    test_data <- full_data[-sample, ]
+    
+    # Combine each ensemble member with all predictors
+    models <- list()
+    maps <- list()
+    predictions <- list()
+    ensembles <- list()
+    for (i in seq_along(carbon_data)) {
```

Comment:
Doesn't need any change for this PR, but making a note for those doing future development that this would be a great place to either parallelize this work or to modify the ML model to take time into account explicitly.